### PR TITLE
server: drop locks and auth pins before waiting for trunc

### PR DESCRIPTION
**Pull request by xiexingguo**
Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

File: src/mds/Server.cc [hunk 1]
```diff
@@ -3750,6 +3750,8 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
     if (truncating_smaller && pi->is_truncating()) {
       dout(10) << " waiting for pending truncate from " << pi->truncate_from
 	       << " to " << pi->truncate_size << " to complete on " << *cur << dendl;
+      mds->locker->drop_locks(mdr.get());
```
Comment (jcsp) [inline src/mds/Server.cc]: This doesn't seem unreasonable, but why do it here and not in the other place we use WAIT_TRUNC above?

**Comment by ukernel** [comment]
looks good

**Comment by xiexingguo** [comment]
@ukernel Thanks!

**Comment by xiexingguo** [comment]
@jcsp Thanks for the testing and merging:-)

**Comment by gregsfortytwo** [comment]
I guess I missed the window, but can somebody explain why it's okay to drop locks here? Amongst other things, we've already succeeded in locking and the mdr remembers that, so it won't try to acquire the locks again on the next go-round...

**Comment by gregsfortytwo** [comment]
Oh, I guess we do clear out the done_locking flag. So it's just a question of whether we're okay re-ordering ops. I don't really think we are since we've already started removing the file data, but we haven't changed the truncate sizes yet?

**Comment by gregsfortytwo** [comment]
Sorry for the noise, happy now.
https://github.com/ceph/ceph/pull/9716

**Pull request merged**